pageserver/src/storage_sync2.rs

```diff
@@ -659,15 +668,61 @@ impl RemoteTimelineClient {
         Ok(())
     }
 
+    /// Schedule a batch of layer file uploads with subsequent IndexPart upload.
+    /// No-op if layer_files is empty.
+    /// If an error is returned, it is undefined which upload operations were scheduled
+    /// and wich ones weren't.
```

Comment:
Looking at the callsites, I think this is unchanged. However we don't have to have UB in our codebase, it could be highlighted what really happens ... Well, I guess this could be restructured to first check that all those are in place.

```rust
// 1. split out the checks
fn verify_uploaded(_: &LayerFileName, layer_metadata: &LayerFileMetadata) -> anyhow::Result<()> {
    // The file size can be missing for files that were created before we tracked that
    // in the metadata, but it should be present for any new files we create.
    ensure!(
        layer_metadata.file_size().is_some(),
        "file size not initialized in metadata"
    );
    
    Ok(())
}

// 2. call verify_uploaded *schedule_layer_file_upload*
// 3. make schedule_layer_file_upload_impl infallible
// 4.in schedule_layer_file_uploads:

pub fn schedule_layer_file_uploads(
    ...
) -> anyhow::Result<()> {
    if layer_files.is_empty() {
        return Ok(());
    }
    
    layer_files.iter().try_for_each(verify_uploaded)?;
    
    let mut guard = self.upload_queue.lock().unwrap();
    let uq = guard.initialized_mut()?;
    
    for (name, md) in layer_files {
        self.schedule_layer_file_upload_impl(uq, name, md); // <-- infallible
    }

    // rest cannot fail either, not reproduced here.
}
```

Reply:
I like this proposal, will implement it.

Reply:
Nice observation, implemented it!